text/0000-imply-option.md

```diff
@@ -0,0 +1,139 @@
+- Feature Name: imply-option
+- Start Date: 2017-10-18
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is an RFC to reduce common boiler plate code when making use of the `Option` type. Similar in intention and motivation to the `Try` trait for `Result`.
+
+# Motivation
+[motivation]: #motivation
+
+This addition will increase the legibility of code segments and assist in defining the thought processes and motivations of programmers through code. The use cases of this addition are solutions which are expressable in the following predicate form:
+```
+    P(x) : Predicate on `x`.
+    F(y) : Function on `y`
+    P(x) -> F(y)
+```
+Or the following Rust psudocode:
+```
+    if P(x) {
+        Some(F(y))
+    } else {
+        None
+    }
+```
+The outcome of this addition will reduce repeated code which introduces bugs during refactoring and present the thought process of the programmer in a clearer fashion through their code.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Option` type is useful when your code may or may not yield a return value.
+Such code may looks similar to this:
+```
+    let x = 0;
+    
+    if x == 0 {
+        Some(x)
+    } else {
+        None
+    }
+```
+However only the `if` branch of this code segment is the important part we're concerned about in our code:
+```
+    if x == 0 {
+        Some(x)
+    }
+```
+But the `else` branch is required for returning `None` value if `x == 0` evaluates to false.
+Fortunately Rusts `Option` type has functionality to get rid of the unecessary code:
+```
+    let x = 0;
+    
+    Option::on_pred(x == 0, x)
+```
+This code has the exact same behaviour as our original `if` statement. Our code is however compressed to a single line and our intentions are just as clear.
+Have you spotted the possible issue with this solution introduces however? What about this code:
+```
+    Option::on_pred(false, foo())
+```
+The above line of code will always return `None` and always throw away the result of `foo()` wasting our precious computing power every time our code needs to return `None`.
+Rust has thought ahead of this problem though:
```

Comment:
This still suggests rust is perfect already and already includes this fancy method below which solves everything. If the proposal for this RFC is to add 2 methods, please state that in the summary with a single line description for each.

Reply:
Again that is the format which is requested in the RFC template of `rust-lang/rfcs`.